packages/kbn-data-stream-adapter/src/data_stream_adapter.ts

```diff
@@ -0,0 +1,157 @@
+import { createOrUpdateDataStream } from './create_or_update_data_stream';
+import { createOrUpdateIndexTemplate } from './create_or_update_index_template';
+import { InstallShutdownError, installWithTimeout } from './install_with_timeout';
+import { getComponentTemplate, getIndexTemplate } from './resource_installer_utils';
+
+export interface DataStreamAdapterParams {
+  kibanaVersion: string;
+  totalFieldsLimit?: number;
+}
+export interface SetComponentTemplateParams {
+  name: string;
+  fieldMap: FieldMap;
+  settings?: IndicesIndexSettings;
+  dynamic?: 'strict' | boolean;
+}
+export interface SetIndexTemplateParams {
+  name: string;
+  componentTemplateRefs?: string[];
+  namespace?: string;
+  template?: IndicesPutIndexTemplateIndexTemplateMapping;
+  hidden?: boolean;
+}
+
+export interface GetInstallFnParams {
+  logger: Logger;
+  pluginStop$: Subject<void>;
+  tasksTimeoutMs?: number;
+}
+export interface InstallParams {
+  logger: Logger;
+  esClient: ElasticsearchClient;
+  pluginStop$: Subject<void>;
+  tasksTimeoutMs?: number;
+}
+
+const DEFAULT_FIELDS_LIMIT = 2500;
+
+export class DataStreamAdapter {
+  protected readonly kibanaVersion: string;
+  protected readonly totalFieldsLimit: number;
+  protected componentTemplates: ClusterPutComponentTemplateRequest[] = [];
+  protected indexTemplates: IndicesPutIndexTemplateRequest[] = [];
+  protected installed: boolean;
+
+  constructor(protected readonly name: string, options: DataStreamAdapterParams) {
+    this.installed = false;
+    this.kibanaVersion = options.kibanaVersion;
+    this.totalFieldsLimit = options.totalFieldsLimit ?? DEFAULT_FIELDS_LIMIT;
+  }
+
+  public setComponentTemplate(params: SetComponentTemplateParams) {
+    if (this.installed) {
+      throw new Error('Cannot set component template after install');
+    }
+    this.componentTemplates.push(getComponentTemplate(params));
+  }
+
+  public setIndexTemplate(params: SetIndexTemplateParams) {
+    if (this.installed) {
+      throw new Error('Cannot set index template after install');
+    }
+    this.indexTemplates.push(
+      getIndexTemplate({
+        ...params,
+        indexPatterns: [this.name],
+        kibanaVersion: this.kibanaVersion,
+        totalFieldsLimit: this.totalFieldsLimit,
+      })
+    );
+  }
+
+  protected getInstallFn({ logger, pluginStop$, tasksTimeoutMs }: GetInstallFnParams) {
+    return async (promise: Promise<void>, description?: string): Promise<void> => {
+      try {
+        await installWithTimeout({
+          installFn: () => promise,
+          description,
+          timeoutMs: tasksTimeoutMs,
+          pluginStop$,
+        });
+      } catch (err) {
+        if (err instanceof InstallShutdownError) {
+          logger.info(err.message);
+        } else {
+          throw err;
+        }
+      }
+    };
+  }
+
+  public async install({ logger, esClient, pluginStop$, tasksTimeoutMs }: InstallParams) {
```

Comment:
It will be very helpful if install method will return the status of the installation. Something like:

```
export interface InitializationPromise {
  result: boolean;
  error?: string;
}
```
We also can add the retry logic here.

Reply:
reference in my PR 

 https://github.com/elastic/kibana/pull/173487/files#diff-fbd58b0d5360e94be1dcf0e98b4fd6465ac7299e2e7473f4cf43f5b852e57f93R104-R116

Reply:
Hey Yuliia, not a bad idea, but I think it does not add any value. The `install` function already throws an error if something goes wrong. And the result will always be `{ acknowledged: true }`, unless there's an error, so there's no need to check it, if it does not fail it means everything went fine.
So we can simplify to this:

```
try {
  await install();
  // success
} catch (err) {
  // error
}

or 

install()
.then() // success
.catch(); // error
```

I don't think we need to add anything else. 
Also, the `install` function already retries every step individually using `retryTransientEsErrors` internally, so we don't have to do it outside 👍